### ISSUE TITLE
Add ability for EdgeApplications to target NodeGroups by label

### ISSUE THREAD

**Issue by siredmar** [kubeedge/kubeedge#3959]
**What would you like to be added/modified**:

I'd like to add the ability to use NodeGroups not only by name, but also by label. 

**Why is this needed**:
Using the scenario below it is possible to handle an EdgeApplication for different (dynamic created/deleted) NodeGroups.

In the example below there are two NodeGroups: blue and green containing blue and green nodes accordingly.
Both NodeGroups share the same label `env: dev`.
The EdgeApplication applies to all NodeGroups that match the label `env: dev`. If a new matching NodeGroups appears, it is automatically taken into account and vice versa if the NodeGroup is being removed.  It shall be possible to either use the `name` or `matchLabels` fields in `targetNodeGroups` for each entry independently.

```
apiVersion: groupmanagement.kubeedge.io/v1alpha1
kind: NodeGroup
metadata:
  name: blue
  labels:
    env: dev
spec:
  matchLabels:
    env: blue
---
apiVersion: groupmanagement.kubeedge.io/v1alpha1
kind: NodeGroup
metadata:
  name: green
  labels:
    env: dev
spec:
  matchLabels:
    env: green
---
apiVersion: groupmanagement.kubeedge.io/v1alpha1
kind: EdgeApplication
metadata:
  name: nginx-app 
spec:
  workloadTemplate:
    manifests:
    - apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: nginx
    spec:
      selector:
        matchLabels:
          app: nginx
      template:
        metadata:
          labels:
            app: nginx
        spec:
          containers:
          - name: nginx
            image: nginx:latest
          ports:
          - containerPort: 80
  workloadScopes:
    targetNodeGroups:
      - matchLabels: 
            env: dev
        overriders:
          replicas: 2
          imageOverrider:
            - component: "registry"
              operator: "replace"
              value: "dev.registry.io"
```

**Comment by vincentgoat** [comment]
That would be great, looking forward to it.
cc @Congrool

**Comment by Congrool** [comment]
Thanks for your feedback. Sounds good to me. I'll implement it later. It's also welcome if you would like to take over this work. @siredmar

**Comment by siredmar** [comment]
@Congrool would be nice to if you'd implement this. I currently have no spare time to do this.

**Comment by Congrool** [comment]
@vincentgoat @fisherxu @siredmar 
Hi, I will work on this feature but I am not familiar with the API release process. Could I just modify the API field and make it still stay at v1alpha1.
Additionally, I think it will be better to add some necessary e2e test for group management before adding any new features.

**Comment by vincentgoat** [comment]
@Congrool The added features will not affect compatibility and can be modified.

**Comment by Congrool** [comment]
Progress udpate:
tests for controller manager: #4256

**Comment by siredmar** [comment]
There is a question that needed to be resolved:
Looking at this example, there is an issue when there is a nodeGroup that has both labels `dev1` and `dev2`. 

```yaml
workloadScopes:
    targetNodeGroups:
      - matchLabels: 
            env: dev1
        overriders:
          replicas: 2
      - matchLabels: 
            env: dev2
        overriders:
          replicas: 3
```
Selecting the correct override is basically impossible.  So, as discussed with @Congrool today we decided that in such a case there will be an Event in the edgeApplication that tells the user that there has been an error rolling out this edgeApplication to the specific nodeGroups that match more than one labelSelector. 
Example event:

```
Event: Cannot roll out to nodeGroup <name of nodegroup> because of multiple matchLabels: <list of matching labels>
```

I'll take a look into it and try to implement this.

**Comment by Congrool** [comment]
@batthebee Welcome!

I have a question:

> in a large number of event entries.

What are these large number of event entries? Could you please give some examples?

The original idea of using 

```
Event: Cannot roll out to nodeGroup <name of nodegroup> because of multiple matchLabels: <list of matching labels>
```

is that we need to provide information for users to check why this edgeapplication cannot be deployed at the nodegroup. With \<name of nodegroup\> and \<list of matching labels\>, users can check the `targetNodeGroups` of this edgeappliction to find the conflict and then solve it.

**Comment by Congrool** [comment]
Ohh, seems a new problem will be introduced when using event like
```
Event: Cannot roll out to nodeGroup <name of nodegroup> because matchLabels conflict, matching <case ordinal>, <case ordinal>...
```

If we update the edgeapplication, which results in the change of the order of elems in targetNodeGroups, stale events will make users confused. 🤔

**Comment by batthebee** [comment]
We've been doing some thinking and we think there's a bigger problem behind it. We believe that the concept of overrides is not sustainable.

If you look a little bit into the future, the requirements for overrides will probably increase significantly.

The NodeGroups represent the different environments in which my applications run. Depending on the environment, many other environmental conditions can be added in addition to deployments. 

Let's take daemonsets for example. Daemonsets have no replica-count at all. What happens to this parameter then?
Or we have an external database running locally, but it is running on a different port in environment one than in environment two. This would require overrides for environments vars. etc.

What I want to say is that overrides in the current form will quickly reach their limits and at some point you will end up with templates or patches.

For templating, however, there are already some good solutions that solve this either client side (helm ytt etc.) or server side (kubevela, crossplane, cue).

Long speech, short sense:

We would suggest not implementing overrides for the NodeGroupSelector for now. The different environments can be solved externally by several EdgeApplications, which are generated with Helm or similar.
By the NodeGroupSelector's this can happen also independently of the node group assignment. 
 
This also makes implementation much easier. 

What do you think?

**Comment by batthebee** [comment]
@Congrool `matchLabels` will it be a map with multible labels. And i want to implement a "all-needs-to-match" logic on that. Fine?

**Comment by Congrool** [comment]
> What I want to say is that overrides in the current form will quickly reach their limits and at some point you will end up with templates or patches.

Yeah, I have to say you're right. The reason that we currently only support overriding image and replicas is just for simplicity. We proposed to implement something like this(BTW, it's struct from [karmada](https://github.com/karmada-io/karmada/blob/master/pkg/apis/policy/v1alpha1/override_types.go#:~:text=//%20PlaintextOverrider%20is%20a,%7D)) for generic patch and also provide replica overrider and image overrider as a short cut for users.
```
// PlaintextOverrider is a simple overrider that overrides target fields
// according to path, operator and value.
type PlaintextOverrider struct {
	// Path indicates the path of target field
	Path string `json:"path"`
	// Operator indicates the operation on target field.
	// Available operators are: add, replace and remove.
	// +kubebuilder:validation:Enum=add;remove;replace
	Operator OverriderOperator `json:"operator"`
	// Value to be applied to target field.
	// Must be empty when operator is Remove.
	// +optional
	Value apiextensionsv1.JSON `json:"value,omitempty"`
}
```
But finally we decided to only support overriding replicas and image, because we thought they are what users most need and most frequently use. And it would make the api simpler and impelmentation more easier. We can add something like `PlaintextOverrider` if users acutally encouter the scenario that need override env or something else in future.


> The different environments can be solved externally by several EdgeApplications, which are generated with Helm or similar.

The motivation of EdgeApplication it's to make the lifecycle management of application across many edge sites more easier, including create, update, delete and status check. We assume that all of these instances at differentes edge sites belongs to a abstract concept, which is called EdgeApplication. Then users can manage the EdgeApplication to manage all of these instances overcoming the environment differences. If using serval EdgeApplications, it may break the idea of this feature.

**Comment by Congrool** [comment]
> @Congrool `matchLabels` will it be a map with multible labels. And i want to implement a "all-needs-to-match" logic on that. Fine?

Sounds good to me, but we have to solve the problem of readability. I think we may need a way to identify the overrider to make the event more explicit.

**Comment by siredmar** [comment]
> > @Congrool `matchLabels` will it be a map with multible labels. And i want to implement a "all-needs-to-match" logic on that. Fine?
> 
> Sounds good to me. And how do you think about the readability problem?

I understand the initial idea of the overriders and the overriders. The overriders are basically one step closer to the need of having a templating engine builtin - like you suggested with your PlaintextOverrider. 
From an architectural point of view the whole idea of creating some sort of templating within a resource isn't the best decision.

Leaving The edgeApplication as is, targeting some nodeGroups and skipping the whole overriders part would mean, that you'd create several edgeApplications each individually parameterized for your edge site. 

So, i understand the initial idea of the EdgeApplication. But the complete system will blow up, once you decide to implement a patching system try to use several more Selector types. 

@batthebee is right when he says, you cloud create the several edge applications using the template engine you want. 
If kubeedge wants the functionality of edgeApplications parameterized by some overriders or a real patching system, it should introduce a new CRD that implements the templating resulting in several edge applications. 

This keeps the edgeApplication logic small and maintainable, but keeps the functionality as is to have one resource that controls edgeApplications on different edge sites with different configurations.

So as a first step we'd implement the `matchLabels` feature without taking overriders into account. So there will be no conflicts regarding overriders. So there isn't anything to worry about. 

What do you think?

**Comment by Congrool** [comment]
To ensure I do not miss the point. I'll first repeat it from the view of architecture:

The portion of override in EdgeApplication will be extracted as a new CRD. And several override CRDs will applyed to the EdgeApplication resulting in several EdgeApplicationInstances. These generated EdgeApplicationInstances will finally be deployed to different nodegroups.
Or use something like Helm to generate different EdgeApplications at the very fisrt. Then users can just deploy these generated EdgeApplications.

Right?

**Comment by siredmar** [comment]
I don't know yet how this "templating CRD" would generate the actual EdgeApplications. 
I thinkfor in-cluster templating maybe something like https://github.com/flanksource/template-operator would be already suitable. 
You define a Template (kind Template, apiVersion: templating.flanksource.com/v1) for an EdgeApplication (with both fixed values and variables), and then you define multiple EdgeApplicationOverrider CRDs (you need to create this very small CRD) which acts as the values for the template. Then the template-operator will generate multiple EdgeApplications based on your EdgeApplicationOverrider resources. That's it. 

This means that the complete overriders would be dropped from EdgeApplication CRD enabling more Workload definitions like Daemonset, Statefulset, Job, whatever with no cost at all.

So to summarize: i'd skip this complete overriders thing and leave it to the user how to generate their EdgeApplications or provide template-operator to do the templating. 
This would be the *most* clean implementation.

**Comment by batthebee** [comment]
@Congrool 
> To ensure I do not miss the point. I'll first repeat it from the view of architecture:
> 
> The portion of override in EdgeApplication will be extracted as a new CRD. And several override CRDs will applyed to the EdgeApplication resulting in several EdgeApplicationInstances. These generated EdgeApplicationInstances will finally be deployed to different nodegroups. Or use something like Helm to generate different EdgeApplications for different nodegroups at the very fisrt. Then users can just deploy these generated EdgeApplications.
> 
> Right?

I meant the second option, i.e. that the user can create multiple EdgeApplications with tools like helm, crossplane, kubevela etc..

These are all very powerful and can address almost any requirement.

Of course, it also has its charm to integrate this, as happened, in kubeedge. The main advantage I see is that I can ask at a central point whether my application has been rolled out. 

But with cd-tooling such as argocd or fluxcd, there are already powerful tools for this.

Since we will soon have the option of selecting nodes by labels, the combination of helm and argocd, for example, can be used to achieve everything that is necessary without having to do a lot of development work within kubeedge.

But maybe I'm missing something. What do you think?

**Comment by Congrool** [comment]
Ok, I got it. Using template engine like helm is a brand new way to overcoming the environment differences of different nodegroups and is attractive to me. And the option 1 which uses EdgeApplicationOverrider CRD will also introduce huge changes into current architecture. We may need more comprehensive consideration to trade off among the original implementation and these two proposals before making the decision. But I currently do not have so much time for POC. Anyone else having the interest in advancing the work is also welcome. @vincentgoat @fisherxu Do you have any suggestions? 

> So to summarize: i'd skip this complete overriders thing and leave it to the user how to generate their EdgeApplications or provide template-operator to do the templating.

Maybe we will have a new architecture in the future. But currently we need to implement this feature based on the current architecture. If we do not consider the override part, it will still have undeterministic side effects and users will know nothing about it which I think is unbearable.

From my perspective, we can currently follow the original proposal to make it work around based on the current archtecture and put the improvement work in the future.

> Selecting the correct override is basically impossible. So, as discussed with @Congrool today we decided that in such a case there will be an Event in the edgeApplication that tells the user that there has been an error rolling out this edgeApplication to the specific nodeGroups that match more than one labelSelector.

What do you think? @siredmar @batthebee

**Comment by Congrool** [comment]
BTW, could you please also describe your use case and motifivation of why you need this feature. It will help others to understand the background.

**Comment by batthebee** [comment]
Motivation:

We have many independently operated nodes on which the same software stack, consisting of deployments, daemonsets and services, is to be rolled out. It is important that the corresponding services are only accessible locally on the devices.

This is currently realized by giving each individual node its own NodeGroup and manually entering all NodeGroups in the edge application. Accordingly, we always have a manual step as soon as a new NodeGroup is added.

Extending the EdgeApplication with a label selector, we can label the new NodeGroup and automatically receive the software stack without manual entry into the EdgeApplication.

**Comment by batthebee** [comment]
Can you take a first look at my PR? 

The overriders are not yet implemented for the NodeGroupSelector. Is it ok for you to remove them from the crd for the NodeGroupSelector for now?

**Comment by Congrool** [comment]
> Can you take a first look at my PR?
> 
> The overriders are not yet implemented for the NodeGroupSelector. Is it ok for you to remove them from the crd for the NodeGroupSelector for now?

Sorry, I'm a little busy these days, I'll take a look soon.

**Comment by siredmar** [comment]
@batthebee @Congrool i tested the controller-manager with the modifications. It's working like a charm. 
I tested every combination i could think of, e.g. multiple nodes, multiple edgeApplications, defining multiple labelSelectors that OR for different groups and multiple labels within a group that must AND together. Removing nodeGroups, labeling new nodeGroups, ...
